core/trino-main/src/main/java/io/trino/execution/scheduler/SqlQueryScheduler.java

```diff
@@ -237,605 +227,1223 @@ private SqlQueryScheduler(
+        public void taskCreated(PlanFragmentId fragmentId, RemoteTask task)
+        {
+            Map<TaskId, URI> bufferLocations = ImmutableMap.of(
+                    task.getTaskId(),
+                    uriBuilderFrom(task.getTaskStatus().getSelf())
+                            .appendPath("results")
+                            .appendPath("0").build());
+            queryStateMachine.updateOutputLocations(bufferLocations, false);
         }
-        else {
-            NodePartitionMap nodePartitionMap = partitioningCache.apply(partitioningHandle);
-            List<InternalNode> partitionToNode = nodePartitionMap.getPartitionToNode();
-            // todo this should asynchronously wait a standard timeout period before failing
-            checkCondition(!partitionToNode.isEmpty(), NO_NODES_AVAILABLE, "No worker nodes available");
-            return Optional.of(nodePartitionMap.getBucketToPartition());
+
+        @Override
+        public void noMoreTasks(PlanFragmentId fragmentId)
+        {
+            queryStateMachine.updateOutputLocations(ImmutableMap.of(), true);
         }
     }
 
-    private static class PipelinedScheduler
+    private static class CoordinatorStagesScheduler
     {
+        private static final int[] SINGLE_PARTITION = new int[] {0};
+
         private final QueryStateMachine queryStateMachine;
-        private final ExecutionSchedule executionSchedule;
-        private final Map<StageId, StageScheduler> stageSchedulers;
-        private final SplitSchedulerStats schedulerStats;
+        private final NodeScheduler nodeScheduler;
+        private final Map<PlanFragmentId, OutputBufferManager> outputBuffersForStagesConsumedByCoordinator;
+        private final Map<PlanFragmentId, Optional<int[]>> bucketToPartitionForStagesConsumedByCoordinator;
+        private final TaskLifecycleListener taskLifecycleListener;
+        private final StageManager stageManager;
         private final List<PipelinedStageExecution> stageExecutions;
+        private final DynamicFilterService dynamicFilterService;
+        private final AtomicReference<DistributedStagesScheduler> distributedStagesScheduler;
+        private final TaskManager coordinatorTaskManager;
 
-        private final AtomicBoolean started = new AtomicBoolean();
+        private final AtomicBoolean scheduled = new AtomicBoolean();
 
-        private PipelinedScheduler(
+        public static CoordinatorStagesScheduler create(
                 QueryStateMachine queryStateMachine,
-                ExecutionSchedule executionSchedule,
-                Map<StageId, StageScheduler> stageSchedulers,
-                SplitSchedulerStats schedulerStats,
-                List<PipelinedStageExecution> stageExecutions)
+                NodeScheduler nodeScheduler,
+                StageManager stageManager,
+                FailureDetector failureDetector,
+                Executor executor,
+                DynamicFilterService dynamicFilterService,
```

Comment:
Why do we need DynamicFilterService in CoordinatorStagesScheduler ?
Would dynamic filtering ever apply to COORDINATOR_DISTRIBUTION stages ?

Reply:
That's a good point. I assume it wouldn't. Let me remove it.